File: src/main/java/dev/stocky37/epic7/core/LevelledHero.java
package dev.stocky37.epic7.core;

import com.google.common.collect.Iterables;
import dev.stocky37.epic7.json.Hero;
import org.immutables.value.Value;

@Value.Immutable
public abstract class LevelledHero {
	abstract Hero hero();

	abstract int stars();

	abstract int level();

	abstract int awakening();

	public Stats getAwakenedStats() {
		return getBaseStats().apply(getAwakenings());
	}

	public Stats getBaseStats() {
		switch(stars()) {
			case 5:
				return hero().getLvl50Stats();
			case 6:
				return hero().getLvl60Stats();
			default:
				throw new IllegalStateException();
		}
	}

	public Iterable<Stats> getAwakenings() {
		return Iterables.limit(hero().getAwakeningStats(), awakening());
	}
}

File: src/main/kotlin/dev/stocky37/epic7/core/LevelledHero.kt
package dev.stocky37.epic7.core

import dev.stocky37.epic7.json.Hero

data class LevelledHero(val hero: Hero, val level: Int, val stars: Int, val awakening: Int) {

	val baseStats: Stats by lazy {
		when(stars) {
			5 -> hero.lv50Stats
			6 -> hero.lv60Stats
			else -> throw IllegalStateException()
		}
	}

	val awakenedBaseStats: Stats by lazy {
		baseStats.apply(Stats.from(*awakenings.toTypedArray()))
	}

	val awakenings: List<Stats> by lazy {
		hero.awakenings.slice(0 until awakening)
	}

}